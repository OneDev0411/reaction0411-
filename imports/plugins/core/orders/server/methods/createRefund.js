import { check } from "meteor/check";
import Hooks from "@reactioncommerce/hooks";
import Logger from "@reactioncommerce/logger";
import ReactionError from "@reactioncommerce/reaction-error";
import { Orders } from "/lib/collections";
import Reaction from "/imports/plugins/core/core/server/Reaction";
import getGraphQLContextInMeteorMethod from "/imports/plugins/core/graphql/server/getGraphQLContextInMeteorMethod";
import { getPaymentMethodConfigByName } from "/imports/plugins/core/payments/server/no-meteor/registration";
import sendOrderEmail from "../util/sendOrderEmail";

/**
 * @name orders/refunds/create
 * @method
 * @memberof Orders/Methods
 * @summary Apply a refund to an already captured order
 * @param {String} orderId - order object
 * @param {String} paymentId - ID of payment to refund
 * @param {Number} amount - Amount of the refund, as a positive number
 * @return {null} no return value
 */
export default function createRefund(orderId, paymentId, amount) {
  check(orderId, String);
  check(paymentId, String);
  check(amount, Number);

  const order = Orders.findOne({ _id: orderId });
  if (!order) throw new ReactionError("not-found", "Order not found");

  const authUserId = Reaction.getUserId();

  if (!Reaction.hasPermission("orders", authUserId, order.shopId)) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  const payment = (order.payments || []).find((pmt) => pmt._id === paymentId);
  if (!payment) throw new ReactionError("not-found", "Payment not found");

  const { name } = payment;
  const { functions } = getPaymentMethodConfigByName(name);
  const context = Promise.await(getGraphQLContextInMeteorMethod(authUserId));
  const result = Promise.await(functions.createRefund(context, payment, amount));

  if (!result.saved) {
    Logger.fatal("Attempt to refund payment failed", order._id, paymentId, result.error);
    throw new ReactionError("Attempt to refund payment failed", result.error);
  }

  // List refunds to see if we've now refunded the full amount of this payment
  const allRefunds = Promise.await(functions.listRefunds(context, payment));
  const refundTotal = allRefunds.reduce((sum, refund) => sum + refund.amount, 0);
  const isFullyRefunded = refundTotal === payment.amount;

  Orders.update(
    {
      "_id": orderId,
      "payments._id": paymentId
    },
    {
      $set: {
        "payments.$.status": isFullyRefunded ? "refunded" : "partialRefund"
      },
      $push: {
        "payments.$.transactions": result
      }
    }
  );

  Hooks.Events.run("onOrderRefundCreated", orderId);

  // Send email to notify customer of a refund
  sendOrderEmail(order, "refunded");
}
