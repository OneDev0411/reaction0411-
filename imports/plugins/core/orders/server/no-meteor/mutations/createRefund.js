import SimpleSchema from "simpl-schema";
import accounting from "accounting-js";
import Logger from "@reactioncommerce/logger";
import ReactionError from "@reactioncommerce/reaction-error";
import { getPaymentMethodConfigByName } from "/imports/plugins/core/payments/server/no-meteor/registration";
import sendOrderEmail from "../util/sendOrderEmail";

const inputSchema = new SimpleSchema({
  amount: {
    type: Number,
    min: 0.01
  },
  orderId: String,
  paymentId: String,
  reason: {
    type: String,
    optional: true
  }
});

/**
 * @method createRefund
 * @summary Use this mutation to create a refund on an orderpayment
 * @param {Object} context - an object containing the per-request state
 * @param {Object} input - Necessary input. See SimpleSchema
 * @param {Object} input.amount - amount to refund
 * @param {Object} input.orderId - order ID of order where payment was applied
 * @param {Object} input.paymentId - ID of payment to refund
 * @param {Object} input.reason - reason for refund
 * @return {Promise<Object>} Object with `order` property containing the updated order
 */
export default async function createRefund(context, input) {
  inputSchema.validate(input);

  const {
    amount,
    orderId,
    paymentId,
    reason
  } = input;

  const { appEvents, collections, isInternalCall, userHasPermission, userId } = context;
  const { Orders } = collections;

  // First verify that this order actually exists
  const order = await Orders.findOne({ _id: orderId });
  if (!order) throw new ReactionError("not-found", "Order not found");

  // Allow refund for accounts with orders or "order/fulfillment" permissions.
  // When called internally by another plugin, context.isInternalCall
  // can be set to `true` to disable this check.
  if (!isInternalCall && !userHasPermission(["orders", "order/fulfillment"], order.shopId)) {
    throw new ReactionError("access-denied", "Access Denied");
  }

  // Verify payment exists
  const payment = (order.payments || []).find((pmt) => pmt._id === paymentId);
  if (!payment) throw new ReactionError("not-found", "Payment not found");

  // Verify that payment method will accept refunds
  const { name } = payment;
  const { canRefund, functions } = getPaymentMethodConfigByName(name);
  if (!canRefund) throw new ReactionError("invalid", "Refunding not supported");

  // Find total of any previous refunds, and determine how much balance is left
  // on this payment that can be refunded.
  const previousRefunds = await functions.listRefunds(context, payment);
  const previousRefundsTotal = previousRefunds.reduce((sum, refund) => sum + refund.amount, 0);
  const amountNotYetRefunded = payment.amount - previousRefundsTotal;

  // Refund the lower amount between `amountNotYetRefunded`, and the `amount` that was passed
  const amountToRefund = Math.min(amountNotYetRefunded, amount);
  const result = await functions.createRefund(context, payment, amountToRefund, reason);

  if (!result.saved) {
    Logger.fatal("Attempt to refund payment failed", order._id, paymentId, result.error);
    throw new ReactionError("Attempt to refund payment failed", result.error);
  }

  // List refunds to see if we've now refunded the full amount of this payment
  const allRefunds = await functions.listRefunds(context, payment);
  const newRefundsTotal = allRefunds.reduce((sum, refund) => sum + refund.amount, 0);

  // There could be JS math errors so we round to 3 decimal places when comparing
  const isFullyRefunded = accounting.toFixed(newRefundsTotal, 3) === accounting.toFixed(payment.amount, 3);


  const { modifiedCount, value: updatedOrder } = await Orders.findOneAndUpdate(
    {
      "_id": orderId,
      "payments._id": paymentId
    },
    {
      $set: {
        "payments.$.status": isFullyRefunded ? "refunded" : "partialRefund"
      },
      $push: {
        "payments.$.transactions": result
      }
    }
  );

  if (modifiedCount === 0 || !updatedOrder) throw new ReactionError("server-error", "Unable to update order");

  await appEvents.emit("afterOrderUpdate", {
    order: updatedOrder,
    updatedBy: userId
  });

  // Send email to notify customer of a refund
  sendOrderEmail(context, order, "refunded");

  return { order: updatedOrder };
}
