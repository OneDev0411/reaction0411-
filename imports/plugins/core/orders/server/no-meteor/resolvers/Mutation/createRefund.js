import { decodeOrderOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/order";
import { decodePaymentOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/payment";

/**
 * @name Mutation/createRefund
 * @method
 * @memberof Payments/GraphQL
 * @summary resolver for the createRefund GraphQL mutation
 * @param {Object} _ - unused
 * @param {Object} args.input - an object of all mutation arguments that were sent by the client
 * @param {Object} args.input.amount - amount to refund
 * @param {Object} args.input.orderId - order ID of order where payment was applied
 * @param {Object} args.input.paymentId - ID of payment to refund
 * @param {Object} [args.input.reason] - reason for refund
 * @param {Object} context - an object containing the per-request state
 * @returns {Promise<Object>} createRefundPayload
 */
export default async function createRefund(_, { input }, context) {
  const {
    amount,
    clientMutationId = null,
    orderId,
    paymentId,
    reason
  } = input;

  const { order } = await context.mutations.createRefund(context, {
    amount,
    orderId: decodeOrderOpaqueId(orderId),
    paymentId: decodePaymentOpaqueId(paymentId),
    reason
  });

  return {
    clientMutationId,
    order
  };
}
