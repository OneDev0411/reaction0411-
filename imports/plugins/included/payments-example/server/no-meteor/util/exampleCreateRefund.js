/**
 * @name exampleCreateRefund
 * @method
 * @summary Create a refund for an order for example payment method
 * @param {Object} context an object containing the per-request state
 * @param {Object} payment object containing transaction ID
 * @param {Number} amount the amount to be refunded
 * @param {String} [reason] the reason for the refund
 * @returns {Object} refund result
 * @private
 */
export default async function exampleCreateRefund(context, payment, amount, reason) {
  const { currencyCode, transactionId } = payment;
  await context.collections.ExampleIOUPaymentRefunds.insertOne({
    amount,
    createdAt: new Date(),
    currencyCode,
    reason,
    transactionId
  });
  return { saved: true };
}
