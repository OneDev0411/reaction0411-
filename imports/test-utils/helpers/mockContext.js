const mockContext = {
  accountId: "FAKE_ACCOUNT_ID",
  appEvents: {
    emit() {},
    on() {}
  },
  collections: {},
  getAbsoluteUrl: jest.fn().mockName("getAbsoluteUrl").mockImplementation((path) => {
    const adjustedPath = path[0] === "/" ? path : `/${path}`;
    return `https://app.mock${adjustedPath}`;
  }),
  getFunctionsOfType: jest.fn().mockName("getFunctionsOfType").mockReturnValue([]),
  mutations: {},
  queries: {},
  userHasPermission: jest.fn().mockName("userHasPermission"),
  userId: "FAKE_USER_ID"
};

/**
 * @summary Returns a mock collection instance with the given name
 * @param {String} collectionName The collection name
 * @returns {Object} Mock collection instance
 */
export function mockCollection(collectionName) {
  return {
    insert() {
      throw new Error("insert mongo method is deprecated, use insertOne or insertMany");
    },
    remove() {
      throw new Error("remove mongo method is deprecated, use deleteOne or deleteMany");
    },
    update() {
      throw new Error("update mongo method is deprecated, use updateOne or updateMany");
    },
    bulkWrite: jest.fn().mockName(`${collectionName}.bulkWrite`).mockReturnValue(Promise.resolve({
      nMatched: 2,
      nModified: 2,
      result: {
        writeErrors: []
      }
    })),
    deleteOne: jest.fn().mockName(`${collectionName}.deleteOne`).mockReturnValue(Promise.resolve({
      deletedCount: 1
    })),
    deleteMany: jest.fn().mockName(`${collectionName}.deleteMany`),
    find: jest
      .fn()
      .mockName(`${collectionName}.find`)
      .mockReturnThis(),
    findOne: jest.fn().mockName(`${collectionName}.findOne`),
    findOneAndDelete: jest.fn().mockName(`${collectionName}.findOneAndDelete`),
    findOneAndUpdate: jest.fn().mockName(`${collectionName}.findOneAndUpdate`),
    fetch: jest.fn().mockName(`${collectionName}.fetch`),
    insertOne: jest.fn().mockName(`${collectionName}.insertOne`),
    insertMany: jest.fn().mockName(`${collectionName}.insertMany`),
    toArray: jest.fn().mockName(`${collectionName}.toArray`),
    updateOne: jest.fn().mockName(`${collectionName}.updateOne`).mockReturnValue(Promise.resolve({
      matchedCount: 1,
      modifiedCount: 1
    })),
    updateMany: jest.fn().mockName(`${collectionName}.updateMany`)
  };
}

[
  "Accounts",
  "Assets",
  "Cart",
  "Catalog",
  "Emails",
  "ExampleIOUPaymentRefunds",
  "Groups",
  "MediaRecords",
  "NavigationItems",
  "NavigationTrees",
  "Notifications",
  "Orders",
  "Packages",
  "Products",
  "Revisions",
  "roles",
  "SellerShops",
  "Shipping",
  "Shops",
  "SimpleInventory",
  "Tags",
  "Templates",
  "Themes",
  "users"
].forEach((collectionName) => {
  mockContext.collections[collectionName] = mockCollection(collectionName);
});

mockContext.collections.Media = {
  find: jest.fn().mockName("Media.find"),
  findLocal: jest.fn().mockName("Media.findLocal"),
  findOne: jest.fn().mockName("Media.findOne"),
  findOneLocal: jest.fn().mockName("Media.findOneLocal"),
  insert: jest.fn().mockName("Media.insert"),
  update: jest.fn().mockName("Media.update"),
  remove: jest.fn().mockName("Media.remove")
};

export default mockContext;
