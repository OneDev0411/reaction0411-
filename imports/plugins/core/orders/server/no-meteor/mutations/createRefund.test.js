/* eslint camelcase: 0 */
import nock from "nock";
import Factory from "/imports/test-utils/helpers/factory";
import mockContext from "/imports/test-utils/helpers/mockContext";
import exampleCreateRefund from "imports/plugins/included/payments-example/server/no-meteor/util/exampleCreateRefund.js";
import createRefund from "./createRefund";

beforeEach(() => {
  jest.resetAllMocks();
});

const fakeOrder = {
  _id: "order1",
  payments: [
    {
      _id: "payment1"
    },
    {
      _id: "payment2"
    }
  ],
  shipping: [
    Factory.OrderFulfillmentGroup.makeOne({
      items: [
        Factory.OrderItem.makeOne({
          _id: "abc",
          quantity: 1,
          price: {
            amount: 1,
            currencyCode: "USD"
          },
          subtotal: 1,
          workflow: {
            status: "new",
            workflow: ["new"]
          }
        })
      ],
      workflow: {
        status: "new",
        workflow: ["new"]
      }
    })
  ],
  shopId: "SHOP_ID",
  workflow: {
    status: "new",
    workflow: ["new"]
  }
};

test("throws if permission check fails", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(fakeOrder));

  mockContext.userHasPermission.mockReturnValueOnce(false);

  await expect(createRefund(mockContext, {
    amount: 10,
    orderId: "order1",
    paymentId: "payment1",
    reason: "Customer was unsatisfied with purchase"
  })).rejects.toThrowErrorMatchingSnapshot();

  expect(mockContext.userHasPermission).toHaveBeenCalledWith(["orders", "order/fulfillment"], "SHOP_ID");
});

test("skips permission check if context.isInternalCall", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(fakeOrder));

  mockContext.isInternalCall = true;

  await expect(createRefund(mockContext, {
    amount: 10,
    orderId: "order1",
    paymentId: "payment1",
    reason: "Customer was unsatisfied with purchase"
  }));

  delete mockContext.isInternalCall;

  expect(mockContext.userHasPermission).not.toHaveBeenCalled();
});

test("throws if amount isn't supplied", async () => {
  await expect(createRefund(mockContext, {
    amount: null,
    orderId: fakeOrder._id,
    paymentId: fakeOrder.payments[0]._id,
    reason: "Customer was unsatisfied with purchase"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if orderId isn't supplied", async () => {
  await expect(createRefund(mockContext, {
    amount: 10,
    orderId: null,
    paymentId: fakeOrder.payments[0]._id,
    reason: "Customer was unsatisfied with purchase"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if paymentId isn't supplied", async () => {
  await expect(createRefund(mockContext, {
    amount: 10,
    orderId: fakeOrder._id,
    paymentId: null,
    reason: "Customer was unsatisfied with purchase"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if payment doesn't exist", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(fakeOrder));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(createRefund(mockContext, {
    amount: 10,
    orderId: fakeOrder._id,
    paymentId: "payment3",
    reason: "Customer was unsatisfied with purchase"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("throws if amount is less than $0.01", async () => {
  mockContext.collections.Orders.findOne.mockReturnValueOnce(Promise.resolve(fakeOrder));

  mockContext.userHasPermission.mockReturnValueOnce(true);

  await expect(createRefund(mockContext, {
    amount: 0,
    orderId: fakeOrder._id,
    paymentId: "payment3",
    reason: "Customer was unsatisfied with purchase"
  })).rejects.toThrowErrorMatchingSnapshot();
});

test("should call example payment method createRefund with the proper parameters and return saved = true", async () => {
  const paymentMethod = {
    processor: "Example",
    displayName: "IOU from EK",
    paymentPluginName: "example-paymentmethod",
    method: "credit",
    transactionId: "ch_17hZ4wBXXkbZQs3xL5JhlSgS",
    amount: 19.99,
    status: "completed",
    mode: "captured",
    createdAt: new Date(),
    workflow: { status: "new" },
    metadata: {},
    currencyCode: "USD"
  };

  const refundResult = {
    id: "re_17hZzSBXXkbZQs3xgmmEeOci",
    object: "refund",
    amount: 1999,
    balance_transaction: "txn_17hZzSBXXkbZQs3xr6d9YECZ",
    charge: "ch_17hZ4wBXXkbZQs3xL5JhlSgS",
    created: 1456210186,
    currency: "usd",
    metadata: {},
    reason: null,
    receipt_number: null
  };

  const result = await exampleCreateRefund(mockContext, paymentMethod, paymentMethod.amount, "Reason for refund");

  mockContext.collections.ExampleIOUPaymentRefunds.insertOne.mockReturnValueOnce(Promise.resolve(refundResult));

  expect(result.saved).toBe(true);
});
