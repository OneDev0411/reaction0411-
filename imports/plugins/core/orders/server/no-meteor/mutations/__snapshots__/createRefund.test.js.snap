// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws if amount isn't supplied 1`] = `"Amount is required"`;

exports[`throws if orderId isn't supplied 1`] = `"Order ID is required"`;

exports[`throws if payment doesn't exist 1`] = `"Payment not found"`;

exports[`throws if paymentId isn't supplied 1`] = `"Payment ID is required"`;

exports[`throws if permission check fails 1`] = `"Access Denied"`;
